### PR TITLE
RNNTL Autograd Tests

### PR DESCRIPTION
**Pull request by carolineechen**

### CODE REVIEW

File: test/torchaudio_unittest/rnnt/autograd_impl.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        inputs_ = []
+        for i in inputs:
+            if torch.is_tensor(i):
+                i = i.to(dtype=self.dtype, device=self.device)
+                if enable_all_grad:
+                    i.requires_grad = True
+            inputs_.append(i)
+        assert gradcheck(transform, inputs_, eps=1e-06, atol=1e-05)
+
+    def test_rnnt_loss_small(self):
```
[the remark this answers is not in the document]
Comment (vincentqb) [reply test/torchaudio_unittest/rnnt/autograd_impl.py:29]: let's try small epsilon?

[the remark this answers is not in the document]
Comment (vincentqb) [reply test/torchaudio_unittest/rnnt/autograd_impl.py:29]: cc @albanD

[the remark this answers is not in the document]
Comment (albanD) [reply test/torchaudio_unittest/rnnt/autograd_impl.py:29]: When the numerical value goes crazy, it usually is because you try to gradcheck at a point where the function is not continuously differentiable.
So you might want to give different inputs to make sure you're at a point that is.

File: test/torchaudio_unittest/rnnt/autograd_impl.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        inputs = get_data_basic(self.device)
```
Comment (carolineechen) [inline test/torchaudio_unittest/rnnt/autograd_impl.py:29]: @vincentqb 
using an example from warp-transducer instead, I get the following:
**torchaudio rnnt**
`numerical:tensor([[-3.4612e-304],
[  0.0000e+00],
[  0.0000e+00],
[-1.1240e-312],
[-2.3224e-304],
[ 2.3224e-304],
[-3.4612e-304],
[  0.0000e+00],
[-2.3224e-304],
[  0.0000e+00],
[  0.0000e+00],
[  0.0000e+00],
[  0.0000e+00],
[ 2.3224e-304],
[-2.3224e-304],
[  0.0000e+00],
[-3.3710e+205],
[-3.4612e-304],
[-2.3224e-304],
[-2.3224e-304],
[ 2.3224e-304],
[  0.0000e+00],
[ 2.3224e-304],
[  0.0000e+00],
[-2.3224e-304],
[ 2.3224e-304],
[  0.0000e+00],
[-7.9051e-317],
[-1.1388e-304],
[  0.0000e+00]], dtype=torch.float64)`
`analytical:tensor([[0.1000],
[0.6000],
[0.1000],
[0.1000],
[0.1000],
[0.1000],
[0.1000],
[0.6000],
[0.1000],
[0.1000],
[0.1000],
[0.1000],
[0.2000],
[0.8000],
[0.1000],
[0.1000],
[0.6000],
[0.1000],
[0.1000],
[0.1000],
[0.1000],
[0.1000],
[0.2000],
[0.1000],
[0.1000],
[0.7000],
[0.1000],
[0.2000],
[0.1000],
[0.1000]], dtype=torch.float64, grad_fn=<CopySlices>)`

**warp-transducer**
`numerical:tensor([[ 0.0000],
[-0.4768],
[ 0.2384],
[ 0.2384],
[ 0.2384],
[ 0.0000],
[ 0.2384],
[-0.2384],
[ 0.2384],
[ 0.2384],
[-0.2384],
[ 0.2384],
[ 0.2384],
[ 0.2384],
[ 0.2384],
[ 0.0000],
[ 0.0000],
[ 0.0000],
[ 0.0000],
[ 0.0000],
[ 0.0000],
[ 0.0000],
[-0.4768],
[ 0.0000],
[ 0.0000],
[-0.7153],
[ 0.2384],
[ 0.2384],
[ 0.2384],
[ 0.2384]])`
`analytical:tensor([[-0.1312],
[-0.3999],
[ 0.1770],
[ 0.1770],
[ 0.1770],
[-0.1857],
[ 0.1225],
[-0.1817],
[ 0.1225],
[ 0.1225],
[-0.3209],
[ 0.0627],
[ 0.0693],
[ 0.1262],
[ 0.0627],
[ 0.0546],
[-0.2182],
[ 0.0546],
[ 0.0546],
[ 0.0546],
[ 0.1207],
[ 0.1207],
[-0.4830],
[ 0.1207],
[ 0.1207],
[-0.6926],
[ 0.1687],
[ 0.1865],
[ 0.1687],
[ 0.1687]])`

Comment (carolineechen) [reply test/torchaudio_unittest/rnnt/autograd_impl.py:29]: I think the reason for analytical == logits for torchaudio might be the use of float64, since only float16 and float32 should be supported. Using float 32 results in the following instead
`numerical:tensor([[ 1.5268e+05],
[ 1.3338e+04],
[ 4.2081e+02],
[ 8.6069e+01],
[-6.2895e+02],
[ 5.2094e+02],
[ 1.8239e+02],
[ 5.1594e+02],
[-7.3195e+01],
[-1.2159e+01],
[ 4.6015e+01],
[ 3.7432e+01],
[ 3.1948e+01],
[ 2.7657e+01],
[ 1.4544e+02],
[-3.4332e+01],
[-6.1989e+01],
[ 2.6464e+01],
[ 3.8147e+00],
[ 2.6226e+00],
[ 4.7684e-01],
[ 7.1526e-01],
[-3.8624e+01],
[ 8.5831e+00],
[ 4.7684e-01],
[ 9.5367e-01],
[ 9.5367e-01],
[ 9.5367e-01],
[ 9.5367e-01],
[-2.3842e-01]])`
`analytical:tensor([[ 0.2438],
[-0.5317],
[ 0.2438],
[ 0.2438],
[-0.1996],
[ 0.1468],
[ 0.1468],
[-0.2588],
[ 0.1468],
[-0.1816],
[ 0.0762],
[ 0.0762],
[ 0.0762],
[ 0.0762],
[-0.3046],
[ 0.0760],
[-0.3041],
[ 0.0760],
[ 0.0760],
[ 0.0760],
[ 0.1433],
[ 0.1433],
[-0.5734],
[ 0.1433],
[ 0.1433],
[ 0.2318],
[ 0.2318],
[ 0.2318],
[ 0.2318],
[-0.9273]], grad_fn=<CopySlices>)`
Additionally, these numbers match with running gradcheck on the internal version as well

File: test/torchaudio_unittest/rnnt/autograd_impl.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vincentqb) [reply test/torchaudio_unittest/rnnt/autograd_impl.py:28]: offline discussion from @albanD: since float64 is not implemented, the epsilon needs to be changed from the default to something like 1e-3, and tolerance to ~1e-2

[the remark this answers is not in the document]
Comment (carolineechen) [reply test/torchaudio_unittest/rnnt/autograd_impl.py:28]: eps = 1e-2: P414383132
eps = 1e-3: P414382964

File: test/torchaudio_unittest/rnnt/autograd_impl.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        )
+        loss = NumpyTransducerLoss(blank=data["blank"])
+        
+        self.assert_grad(loss, inputs, enable_all_grad=False, nondet_tol=1e-10)
+
+    @parameterized.expand([
+        (get_B1_T10_U3_D4_data, ),
+        (get_B1_T10_U3_D4_data, ),
+        (get_numpy_data_B2_T4_U3_D3, ),
+        (get_numpy_data_B1_T2_U3_D5, ),
```
Comment (carolineechen) [inline test/torchaudio_unittest/rnnt/autograd_impl.py:143]: both the loss and gradients do not match for this set of data. they match for the other 3 cases

File: torchaudio/csrc/rnnt/cpu/compute.cpp [hunk 4]
```diff
@@ -116,6 +116,18 @@ std::tuple<torch::Tensor, c10::optional<torch::Tensor>> compute(
       /*int_size=*/int_workspace.numel());
 
   switch (logits.scalar_type()) {
+      case torch::ScalarType::Double: {
+      Compute</*DTYPE=*/double, /*CAST_DTYPE=*/float>(
```
Comment (vincentqb) [inline torchaudio/csrc/rnnt/cpu/compute.cpp:120]: from offline discussion, we need to confirm that CAST_TYPE controls the return type. if that is the case, we'll need to add a test to validate that the type matches the one requested by the user instead.

Comment (vincentqb) [reply torchaudio/csrc/rnnt/cpu/compute.cpp:120]: in our case here, it means that gradcheck may only still be receiving float32 and not float64. can you verify that?

**Comment by vincentqb** [comment]
great observations! i'm assuming the same old for gpu, is that the case?

**Comment by vincentqb** [comment]
one piece of information to add is the result of those tests for warp-transducer

File: test/torchaudio_unittest/rnnt/autograd_impl.py (hunk 3 above)
Comment (carolineechen) [reply test/torchaudio_unittest/rnnt/autograd_impl.py:143]: Some weird boundary effect?? middle 3 cols are the same, but cols 0 and 4 seem generally swapped with the exception of the final row
```
tensor([4.4957], grad_fn=<_RNNTBackward>) 
tensor([[[[-0.1312, -0.3999,  0.1770,  0.1770,  0.1770],
          [-0.1857,  0.1225, -0.1817,  0.1225,  0.1225],
          [-0.3209,  0.0627,  0.0693,  0.1262,  0.0627]],

         [[ 0.0546, -0.2182,  0.0546,  0.0546,  0.0546],
          [ 0.1207,  0.1207, -0.4830,  0.1207,  0.1207],
          [-0.6926,  0.1687,  0.1865,  0.1687,  0.1687]]]])
tensor(5.0957, grad_fn=<SumBackward0>)
tensor([[[[ 0.1770, -0.3999,  0.1770,  0.1770, -0.1312],
          [ 0.1225,  0.1225, -0.1817,  0.1225, -0.1857],
          [ 0.0627,  0.0627,  0.0693,  0.1262, -0.3209]],

         [[ 0.0546, -0.2182,  0.0546,  0.0546,  0.0546],
          [ 0.1207,  0.1207, -0.4830,  0.1207,  0.1207],
          [ 0.3074,  0.1687,  0.1865,  0.1687, -0.8313]]]])
```

File: torchaudio/csrc/rnnt/cpu/compute.cpp (hunk 4 above)
Comment (carolineechen) [reply torchaudio/csrc/rnnt/cpu/compute.cpp:120]: yea, looks like it's still float32

File: test/torchaudio_unittest/rnnt/autograd_impl.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        loss = NumpyTransducerLoss(blank=data["blank"])
+        
+        self.assert_grad(loss, inputs, enable_all_grad=False, nondet_tol=1e-10)
+
+    @parameterized.expand([
+        (get_B1_T10_U3_D4_data, ),
+        (get_numpy_data_B2_T4_U3_D3, ),
+        # (get_numpy_data_B1_T2_U3_D5, ),
+    ])
+    def test_warp_transducer_gradcheck(self, data_func):
```
Comment (carolineechen) [inline test/torchaudio_unittest/rnnt/autograd_impl.py:126]: the open source warp-transducer passes gradcheck for `get_B1_T10_U3_D4_data` and `get_numpy_data_B2_T4_U3_D3`, and we have also confirmed that the gradients of the torchaudio and warp-transducer implementations match for both these data sets (comparing the .grad attribute after calling backwards). However, torchaudio is failing gradcheck due to `Backward is not reentrant, i.e., running backward with same input and grad_output multiple times gives different values, although analytical gradient matches numerical gradient.The tolerance for nondeterminism was 1e-10.`

@albanD are there any reasons you can think of for this behavior?

File: torchaudio/csrc/rnnt/cpu/compute.cpp (hunk 4 above)
Comment (vincentqb) [reply torchaudio/csrc/rnnt/cpu/compute.cpp:120]: can you confirm what the results are for numpy with float64?
